Scripts/Ulduar: implement Thorim Encounter

joschiwald opened the pull request:
TODO:
- [x] Loot (nh and hm is swapped?)
- [x] Sif's blizzard 
- [x] Known issue: Thorim targets (and turns! while on balcony, this didn't happen before)
- [x] and everything that need to be done before merge

i will merge and squash the commits myself when it is ready to be merged
(pls ignore stuff that should not be there)

Keader commented:
My God, I have no words to describe the joy of fighting the Boss ...
Congratulations for all your work, script working perfectly here :+1:  :+1:

Shauren commented on src/server/scripts/Northrend/Ulduar/Ulduar/ulduar.h:
```diff
@@ -368,7 +399,7 @@ enum UlduarData
     DATA_THORIM_YS,
     DATA_MIMIRON_YS,
-    DATA_ILLUSION,
+    DATA_ILLUSION, // = 46 used by conditions
```
would be better to explicitly set the value then (instead of just leaving a comment)

joschiwald replied:
it could break something if anyone add an enum before this one (multiple enums with same value)
what do you think about `static_assert(DATA_ILLUSION == 46)` in instancescript?

Treeston replied:
How about explicitly setting it to a high value and adjusting conditions accordingly? Seems more adaptable.

sirikfoll replied:
This is no longer used for conditions, the comment could be removed.

ariel- commented:
Fix'd and fix'd

ariel- commented:
Loot fixed, can you detail the point "everything that need to be done before merge"?

sirikfoll commented on src/server/scripts/Northrend/Ulduar/Ulduar/boss_thorim.cpp:
```diff
@@ -97,7 +916,1278 @@ class boss_thorim : public CreatureScript
+
+            void Register() override
+            {
+                OnEffectHitTarget += SpellEffectFn(spell_thorim_arena_leap_SpellScript::HandleScript, EFFECT_0, SPELL_EFFECT_SCRIPT_EFFECT);
```
Should be SPELL_EFFECT_JUMP_DEST

ariel- replied:
Good catch, ty

DjYXA commented:
This pull can be merged now? A long time with any commits and is working fine...

Keader commented:
@DjYXA The last time I spoke to the Joschiwald he said he lacked time for that. It's been over a week ago that he not online at the IRC

ariel- commented:
Rebased this PR due to unrelated build error.

Keader commented:
All Achievs seem to work well. The Single achiev I could not do (I did not understand well what should be done) is "Who Needs Bloodlust?". Someone confirm if it works?

chaodhib commented:
Looks pretty good. Nicely done!

Sif's Blizzard needs more work though. The only area suspecting of being under a blizzard is the outer ring see here: ![wowwiki](http://vignette1.wikia.nocookie.net/wowwiki/images/3/36/Thorimposition.jpg/revision/latest?cb=20090618200120) and also this very good tutorial https://www.youtube.com/watch?v=nkkIJD2kJyM. This forces a very tight positioning in hard mode. Right now, the Blizzard is expanding too close to the center:
![wowscrnshot_081815_034539](https://cloud.githubusercontent.com/assets/6612710/9320759/f1248e40-455d-11e5-8e4e-02d295834eb1.jpg)

and in some directions, it leaves a safe spot near the wall which shouldn't exist:
![wowscrnshot_081815_034528](https://cloud.githubusercontent.com/assets/6612710/9320746/c7cbd2f6-455d-11e5-8b63-c036d817fc36.jpg)

chaodhib commented:
I reconstructed the room using a 2D software and data IG (.wpgps FTW). I also used several videos to determine safe and unsafe locations and even though I don't gurantee a perfect precision, the result is better than nothing. The result is this:
![roomimage](https://cloud.githubusercontent.com/assets/6612710/9424971/8db54b96-4900-11e5-8143-667258a14fad.png)

* North is the entrance to the arena from the center of Ulduar. East is the entrance to the gauntlet.
* the small blue circles are the data points (the limits of the room) taken IG
* In blue= separation between the blizzard area and the rest of the room
* In red= the path the Blizzard NPC should take to get that Blizzard area

 Check your PRs, I will submit the new waypoints

Keader commented:
http://ptr.wowhead.com/quest=13610/thorims-sigil
This Quest dont work
Sigil is never  dropped

Sometimes the Runic Colossus (32872) is becoming no selectable

sirikfoll commented:
Guess that this may solve the loot issue
```sql
UPDATE `gameobject_loot_template` SET `LootMode`='1' WHERE `Entry`=27074 AND `Item` IN (45784,45928,45929,45930,45931,45933);

StormBytePP commented:
@sirikfoll That is not good: sigil should be only dropped in hard mode, and LootMode=1 means default, so it will be always being dropped even in normal mode.
According to 
```c++
enum LootModes
{
    LOOT_MODE_DEFAULT                  = 0x1,
    LOOT_MODE_HARD_MODE_1              = 0x2,
    LOOT_MODE_HARD_MODE_2              = 0x4,
    LOOT_MODE_HARD_MODE_3              = 0x8,
    LOOT_MODE_HARD_MODE_4              = 0x10,
    LOOT_MODE_JUNK_FISH                = 0x8000
};
```
It needs 0x2, 0x4 or 0x8, not sure which one (or all of them)

Keader commented:
@StormBytePP But in this case, normal and hardmode are different objects.Then I would have no problem

@Sirikfoll work for me

sirikfoll commented:
Like @keader said, the GO with this loot is spawned only if the hard mode is active, no need for `LootMode` in this case

obsidean commented:
just wondering when this will be merged with the main branch?

Aokromes commented:
When joschiwald thinks it's ready for it.

obsidean commented:
if i wanted to test this, could you point me to the instruction on how to merge this branch to my copy of the master branch?

DDuarte commented:
```
git checkout -b joschiwald-thorim 3.3.5
git pull https://github.com/joschiwald/TrinityCore.git thorim
```

obsidean commented:
thank you.

digz6666 commented:
How to start the encounter?

chaodhib commented:
pull Joschiwald's thorim branch. Recompile. Execute the new SQL queries. Start the server.

Then there should be mobs in the arena between each other. Kill them and the encounter will start.

digz6666 commented:
Killed the mobs but it's not starting, am I doing something wrong?

Keader commented:
@digz6666 Did you execute the sql querrys? by killing the trashs the fight should start.

digz6666 commented:
@Keader Yes I executed from this pull request.

digz6666 commented:
@Keader ah the players said there's no add near Thorim chamber. Maybe something left to apply?

digz6666 commented:
@Keader now they says seeing the mobs LEL these players ;)

Keader commented:
Need update after (497aa4425bd8a84acb7f5ebcf03692ef96a64c06)

ariel- commented:
A little update on this. Try to test blizzard now. I'm still not sure of what the real targets are, but I have enhanced the "tracks around the room" aspect of the spell.

ariel- commented:
So, the initial approach was right, Blizzard is casted only on the moving trigger npc, the periodic aura triggers and should target a destination implicitly, I don't know but looks like a bug with the effect targeting system, due to both effects having different implicit targets.

Thanks again @Aokromes

chaodhib commented:
I made a representation of the room at scale not long ago. I'm posting it here in case it's useful:

![roomretailb](https://cloud.githubusercontent.com/assets/6612710/10054526/c602fbf2-6231-11e5-893f-20777512d2f1.png)

ariel- commented:
@chaodhib deb27f220349c752e4383c093d09fd93f79e1a21 Is a work around for the spell targeting issue, could you test if Blizzard goes smoothly now please?

chaodhib commented:
I'm not following. What is the spell targeting issue? My understanding of the spell was that Sif casts it on a bunny NPC who is moving the room, thus controlling the movement of the Blizzard. The only thing left was to determine the right way points of this bunny. Did I miss something?

ariel- commented:
The waypoints are from sniff so presumably correct, the spell was to be targeted 13yds. away to the left of the bunny, however, don't know if a glitch in spell system or wrong dbc data, the spell was targeted right where the bunny is. deb27f2 works around this.

chaodhib commented:
Looks much better. Still doesn't look entirely blizz-like unfortunately. A few weeks ago, I collected a few "safe" and "unsafe" points using videos of the encounter on Youtube.

Here is the data I have:

Coord X |  Coord Y | Type of point | Source
----------- | ----------- | ----------------- | ---------
2152,985 | -262,879 | SAFE | "Method vs Thorim Hard Mode", 03:34
2156,880 | -267,381 | NON SAFE | ??? (sorry, forgot to note where I got this point)
2120,279 | -253,416 | SAFE | "Vodka vs Thorim Hard Mode - Kinaesthesia POV " 4:25

Links:
Method: https://www.youtube.com/watch?v=WUJQrQ-v3T0
Vodka: https://www.youtube.com/watch?v=8BTLfOHyMaQ

I tried the first point today. In the video:
![vlcsnap-2015-09-27-13h49m34s860](https://cloud.githubusercontent.com/assets/6612710/10122582/a54ff654-651e-11e5-82ed-026b9f501a64.png)

IG with the current patch:
![wowscrnshot_092715_131942](https://cloud.githubusercontent.com/assets/6612710/10122589/c9651eac-651e-11e5-996b-9ca6ab80ca45.jpg)

Any player standing in that position should be safe from the blizzard. With the current implementation, he is not. So the blizzard is still a few yards too close to the center IMO. Got no clue of what could be the source of the problem :x

Keader commented:
After spending time to 'hard mode' to Sif should disappear?

Treeston commented:
@Keader anecdotally, yes. she finishes controlling him and teleports out.

Keader commented:
@ariel- Kill credit dont work in dead people. (if you die during the meeting, does not receive achievs)

sirikfoll commented:
Suggestion
```sql
-- for credit target dead players
UPDATE `spell_dbc` SET `AttributesEx2`='5' WHERE `Id` = 64985;
-- for hard mode loot in 10
UPDATE `gameobject_loot_template` SET `LootMode`='1' WHERE `Entry`=27074 AND `Item` IN (45784,45928,45929,45930,45931,45933);

ariel- commented:
There is no problems with the loot right now. @sirikfoll

sirikfoll commented:
Hard mode chest from 10man 
```sql
SELECT * FROM  `gameobject_loot_template` WHERE `Entry`=(SELECT `Data1` FROM `gameobject_template` WHERE `entry`=194313);
```
note that there is some items with `LootMode`=2, and the new script doesn't use LootModes anymore, so these items will never drop

sirikfoll commented:
https://github.com/joschiwald/TrinityCore/blob/thorim/src/server/scripts/Northrend/Ulduar/Ulduar/boss_thorim.cpp#L2155
needs to be updated due to changes  in cba122ad94af5081fa748c5d77842dd5205e7b67
`bool OnConditionCheck(Condition const* condition, ConditionSourceInfo& sourceInfo) override`

Keader commented:
@ariel-  After transiting Thorim for phase 2 , if players remain on the platform and generate a good amount of threat with spells in general, Thorim will return to the platform.
Is this intentional ?

Vid: https://www.youtube.com/watch?v=M5Rxpukyq2k&feature=youtu.be

Complete Try: https://www.youtube.com/watch?v=ve7QtB_cdIo&feature=youtu.be

TC hash: 52692c3f9d09dc781add9cfee043d6f6ec80291a

ccrs commented:
do you have mmaps on?

Keader commented:
@ccrs sorry, i tested in a TC not clean, there was some modification causing it

ariel- commented:
So, the spellInfo needed corrections for correct blizzard movement, however as @chaodhib points out this is still not correct according to the retail fight. Maybe it needs range correction too? (ie 15yds  instead of 13yds)

I know cases where this was solved looking for spell attributes on a different client version. I only have wotlk, so could any kind soul check spells 62576 and 62602 in cata/wod spelldbc?

Rushor commented:
wod dbcs:

```sql
ID - 62576 Blizzard
=================================================
Description: A rain of ice moves around the room, dealing $s1 Frost damage every $t1 sec. for $d.

ToolTip: Deals $s1 damage every $t1 sec.
Slows movement speed by $s2%.
Triggred by spell: (62577) Blizzard
=================================================
Category = 0, SpellIconID = 285, activeIconID = 0, SpellVisual = (7718,0)
Family SPELLFAMILY_GENERIC, flag [0] 0x00000000 [1] 0x00000000 [2] 0x00000000 [3] 0x00000000

SpellSchoolMask = 16 (SPELL_SCHOOL_MASK_FROST)
DamageClass = 1 (SPELL_DAMAGE_CLASS_MAGIC)
PreventionType = 1 (SPELL_PREVENTION_TYPE_SILENCE)
=================================================
Attributes: 0x00000800 (SPELL_ATTR0_UNK11)
AttributesEx1: 0x00000088 (SPELL_ATTR1_CANT_BE_REDIRECTED, SPELL_ATTR1_CANT_BE_REFLECTED)
AttributesEx2: 0x00000004 (SPELL_ATTR2_CAN_TARGET_NOT_IN_LOS)
AttributesEx4: 0x00002001 (SPELL_ATTR4_IGNORE_RESISTANCES, SPELL_ATTR4_UNK13)
AttributesEx6: 0x00000400 (SPELL_ATTR6_CAN_TARGET_POSSESSED_FRIENDS)
=================================================
Spell Level = 0, base 0, max 0

Category = 0
DispelType = 0 (DISPEL_NONE)
Mechanic = 0 (MECHANIC_NONE)
SpellRange: (Id 1) "Self Only":
    MinRangeNegative = 0, MinRangePositive = 0
    MaxRangeNegative = 0, MaxRangePositive = 0

Cast time (Id 1) = 0,00
Duration: ID (1)  10000, 0, 10000
Interrupt Flags: 0x00000000, AuraIF 0x00000000, ChannelIF 0x00000000
Chance = 0, charges - 0
=================================================
Effect 0: Id 27 (SPELL_EFFECT_PERSISTENT_AREA_AURA) DIFFICULTY_NONE
BasePoints = 2188 to 2812
Targets (18, 28) (TARGET_DEST_CASTER, TARGET_DEST_DYNOBJ_ENEMY)
Aura Id 3 (SPELL_AURA_PERIODIC_DAMAGE), value = 2187, misc = 0 (0), miscB = 0, periodic = 1000
Radius (Id 17) 13,00

Effect 1: Id 27 (SPELL_EFFECT_PERSISTENT_AREA_AURA) DIFFICULTY_NONE
BasePoints = -65
Targets (50, 28) (TARGET_DEST_CASTER_LEFT, TARGET_DEST_DYNOBJ_ENEMY)
Aura Id 33 (SPELL_AURA_MOD_DECREASE_SPEED), value = -65, misc = 0 (0), miscB = 0, periodic = 0
Radius (Id 17) 13,00
Effect Mechanic = 11 (MECHANIC_SNARE)
```

```sql
ID - 62602 Blizzard
=================================================
Description: Ein Eisregen bewegt sich durch den Raum, verursacht $d lang alle $t1 Sek. $s1 Frostschaden.
ToolTip: Verursacht alle $t1 Sek. $s1 Schaden
Bewegungstempo um $s2% verlangsamt.
Category = 0, SpellIconID = 285, activeIconID = 0, SpellVisual = (7718,0)
Family SPELLFAMILY_GENERIC, flag [0] 0x00000000 [1] 0x00000000 [2] 0x00000000

SpellSchoolMask = 16 (SPELL_SCHOOL_MASK_FROST)
DamageClass = 1 (SPELL_DAMAGE_CLASS_MAGIC)
PreventionType = 1 (SPELL_PREVENTION_TYPE_SILENCE)
=================================================
Attributes: 0x00000800 (SPELL_ATTR0_UNK11)
AttributesEx1: 0x00000088 (SPELL_ATTR1_CANT_BE_REDIRECTED, SPELL_ATTR1_CANT_BE_REFLECTED)
AttributesEx2: 0x00000004 (SPELL_ATTR2_CAN_TARGET_NOT_IN_LOS)
AttributesEx4: 0x00002001 (SPELL_ATTR4_IGNORE_RESISTANCES, SPELL_ATTR4_UNK13)
AttributesEx6: 0x00000400 (SPELL_ATTR6_CAN_TARGET_POSSESSED_FRIENDS)
=================================================
Spell Level = 0, base 0, max 0, maxTarget 0

Category = 0
DispelType = 0 (DISPEL_NONE)
Mechanic = 0 (MECHANIC_NONE)
SpellRange: (Id 1) "Nur selbst":
    MinRange = 0, MinRangeFriendly = 0
    MaxRange = 0, MaxRangeFriendly = 0

CastingTime (Id 1) = 0,00
Duration: ID (1)  10000, 0, 10000
Interrupt Flags: 0x00000000, AuraIF 0x00000000, ChannelIF 0x00000000
Chance = 101, charges - 0
=================================================
Effect 0: Id 27 (SPELL_EFFECT_PERSISTENT_AREA_AURA)
BasePoints = 4813 to 6188
Targets (18, 28) (TARGET_DEST_CASTER, TARGET_DEST_DYNOBJ_ENEMY)
Aura Id 3 (SPELL_AURA_PERIODIC_DAMAGE), value = 4813, misc = 0 (0), miscB = 0, periodic = 1000
Radius (Id 17) 13,00

Effect 1: Id 27 (SPELL_EFFECT_PERSISTENT_AREA_AURA)
BasePoints = -65
Targets (50, 28) (TARGET_DEST_CASTER_LEFT, TARGET_DEST_DYNOBJ_ENEMY)
Aura Id 33 (SPELL_AURA_MOD_DECREASE_SPEED), value = -65, misc = 0 (0), miscB = 0, periodic = 0
Radius (Id 17) 13,00
Effect Mechanic = 11 (MECHANIC_SNARE)

Effect 2:  NO EFFECT
```

ariel- commented:
Thanks Rushor, amazing speed!

I see no differences in the targets nor in the range, so I still suspect there is some problem with the spell targeting system. Either that and/or the Thorim Event Bunny hasn't good sniffed waypoints.

ariel- commented:
Any news here? I just think the whole blizzard issue can be solved with new waypoints for the trigger, I don't know how to use the WP Creator, and more often than not I got an empty list after loading the dump.

chaodhib commented:
If the blizzard is cast on the dummy, the waypoints in this PR can be used: https://github.com/joschiwald/TrinityCore/pull/1

I wrote a small Java program (the code is also in the pull request) to generate these waypoints. I could correct it if you could describe to me what you think is the exact geometry of the waypoints.

It would be good to understand why the waypoints provided by sniff are wrong though.

ariel- commented:
I did not sniff this, the waypoints can be wrong if at any moment the
object is created and deleted (ie player gets out of range); in order to be
correct they should be sniffed with the player at a nearby point, without
moving and waiting for the dummy to complete a round (approximated time
since clientside you can't 'see' the npc patrolling)

2015-12-30 12:39 GMT-03:00 Chaouki Dhib <notifications@github.com>:

> If the blizzard is cast on the dummy, the waypoints in this PR can be
> used: joschiwald#1 <https://github.com/joschiwald/TrinityCore/pull/1>
>
> I wrote a small Java program (the code is also in the pull request) to
> generate these waypoints. I could correct it if you could describe to me
> what you think is the exact geometry of the waypoints.
>
> It would be good to understand why the waypoints provided by sniff are
> wrong though.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/TrinityCore/TrinityCore/pull/15008#issuecomment-168021108>
> .
>

Saben65 commented:
Rush0r has a helpful guide on using WP Creator here. https://community.trinitycore.org/topic/11686-tool-waypoint-creator-how-to-add-waypoints-properly/.

The only thing I would add is that it saves every single WP, so if the creature does his patrol multiple times while you are sniffing, you have to remove the duplicates (select them, and right click -> cut) before generating the SQL. Occasionally the WPs don't stack up perfectly, and the only way I have found to judge if they should be removed or not is your best judgement.

I don't know if the Sniffer will get updated in the next day or so before I have to head back home. But if you still haven't re-sniffed by the end of this week, I will be able to do it for you.

ariel- commented:
I only have some cataclysm-sniffs that show empty WP list when parsing them, if you can do this it'll be great!

chaodhib commented:
So, if I understand correctly, we need a new sniff for this boss? Someone who would go to Thorim's room, kill every mob except one (no need to start the encounter) in order to avoid useless data in the sniff and just stand there, in the middle of the arena, for 2-3 minutes. With a bit of luck, the dummy is doing circles all the time, even if the encounter hasn't started. 

Is that correct? Could someone please do it?

Treeston commented:
@ariel- what npc id are we looking for, I can see if we have any further sniffs with it

sirikfoll commented:
The waypoints are correct, at least for 6.x. in my sniff i got this https://gist.github.com/sirikfoll/3435a93795f6eb3ff0fd (not cleaned up) but it makes the same path as the one from this PR

chaodhib commented:
Sirikfoll's waypoints and the PR's waypoints do indeed describe the same geometry: 
PR waypoints:
![pr wps](https://cloud.githubusercontent.com/assets/6612710/12222306/dec1df14-b7b8-11e5-9ef7-4540ef3e8d47.jpg)
Sirikfoll waypoints:
![sirikfoll wps](https://cloud.githubusercontent.com/assets/6612710/12222307/dec20d04-b7b8-11e5-988d-afcb98a88bee.jpg)

The only thing weird is that, at some random moments, it appears like the NPC crosses the room for no apparent reason (the times where he skip a few waypoints).

Besides that, the geometry is an approximated circle with a 19.4 raduis and a center located in the center of the room. I'll dig in more and see what else I can find.

chaodhib commented:
I have clean up the waypoints. Here is the result (32 waypoints describe a single cycle): http://pastebin.com/2TvF4A2T 

However, it's still not working. My opinion is that the implementation of TARGET_DEST_CASTER_LEFT is bugged.
Before
![wowscrnshot_011216_011529](https://cloud.githubusercontent.com/assets/6612710/12250886/cee3911a-b8ca-11e5-8058-bd5cabf5baf2.jpg)
Then the bunny NPC passes behind me. in other words, i'm on the right side of the bunny (i shouldn't be affected by the blizzard)
![wowscrnshot_011216_011531](https://cloud.githubusercontent.com/assets/6612710/12250892/d395519e-b8ca-11e5-9249-cecafd4d1672.jpg)
and when i disable GM powers;
![wowscrnshot_011216_011534](https://cloud.githubusercontent.com/assets/6612710/12250921/fa21e520-b8ca-11e5-882e-685e78a72368.jpg)

Treeston commented:
So the visual is correct but the core's target selection is incorrect, yes? I can look into it.

tkrokli commented:
Yes, that was the impression I got from chaodib when he last talked with ariel- about it in #trinity.

ariel- commented:
Took a look into this, what's bugged serverside is spell selection code for this case, I think dbc in this case shouldn't be changed and fix the case where you have 2 SPELL_EFFECT_PERSISTENT_AREA_AURA with different targets, so that core sends the right visual and damage is applied properly

Treeston commented:
Elaborate?

chaodhib commented:
@Treeston What do you mean by correct visual? if it affects the correct area? It's hard to tell since the visual of the Blizzard isn't something marked on the ground and easy to see (like Death & Decay for example). However, yes, it *appears* like the visual is correct but the damage zone isn't.

@ariel- The dbc data is wrong for this spell. The correction you applied is needed. It makes absolutely no sense that the area of damage and the area of slow isn't the same. By the way, "Targets (50, 28)  (TARGET_DEST_CASTER_LEFT, TARGET_DEST_DYNOBJ_ENEMY)" is used twice in 3.3.5: for Sif's Blizzard and for the Shade of Aran's Blizzard (Karazhan), spell 29951. For that spell, both spell effects have (50, 28) and it has the exact same behavior we are trying to have here (rotating blizzard in an outer ring).

By talking about this issue on IRC with Aokromes 2 days ago, I'm 99% sure this is a case where the server files used by Blizzard diverges from the client files and we need to override the difference. The code added in SpellMgr.cpp shouldn't be viewed as a workaround but a necessity imo.

In conclusion, I think we should look into how (TARGET_DEST_CASTER_LEFT, TARGET_DEST_DYNOBJ_ENEMY) is handled.

Kittnz commented:
You can use the sniffed waypoints from @sirikfoll

chaodhib commented:
@Kittnz I did. My waypoints are based on sirikfoll's. These sniffed waypoints aren't usable as is for 2 reasons:
* jumps like https://gist.github.com/sirikfoll/3435a93795f6eb3ff0fd#file-thorimblizzbunny-sql-L658 (we can see that the x, y coordinates along with the time stamp jumps. a problem with the sniff I suppose.)
* it's unnecessarily long. I noticed that each cycle is described by only 32 points. The SQL contained 765 WPs so about 24 cycles. There is no benefits to have duplicate cycles so I got rid of them. The **how** I got rid of them is something I could probably improve:

I don't understand why duplicate points aren't located exactly at the same location. I suppose it could be explained by latency and because the server is a concurrency environment, so the npc might not stop at the same location each time.

Anyway, what I did was compute the average of each group of duplicate points (so 32 groups of about 24 points) and I found the average X and the average Y.

Kittnz commented:
Yes these waypoints are never 100% correct, they have a small difference every time they do the path. But all in all it is pretty accurate and it's sniffed data. If @sirikfoll used the waypoint creator, he could have corrected the waypoint easly.

Treeston commented:
Maybe retail server has a concept of "circular path" so the resulting WP sent to client are different each time.

Treeston commented:
So is there incorrect handling of a target type or is it "just" nonsensical dbc data?

chaodhib commented:
Indeed.

Anyway, here is the new script: http://pastebin.com/pnb4DXee
I only used https://gist.github.com/sirikfoll/3435a93795f6eb3ff0fd#file-thorimblizzbunny-sql-L34 to https://gist.github.com/sirikfoll/3435a93795f6eb3ff0fd#file-thorimblizzbunny-sql-L65 of sirikfoll WPs.

The NPC is still flying. Not sure if I messed something up. The speed of which the NPC completes a cycle needs to be worked on (based on the time stamps of the WPs).

chaodhib commented:
@Treeston Imo, both.

Treeston commented:
I'll take a look at some point when I have some free time.

chaodhib commented:
To give a bit more information on the remaining issue. The spell is correctly casted 13y on the left of the NPC bunny (if there is enough place. On the side of the room near Thorim's platform, it will be casted as close to the wall as possible, I'm going to assume it's blizz-like). The issue is that the radius of the Blizzard is bigger then 13 yards. With illustrations, this is what we have at the moment (with exaggeration):
![bunny3](https://cloud.githubusercontent.com/assets/6612710/12297656/71314798-ba0e-11e5-8dfc-657c2f3fc58b.png)
I made some (quick) measurements, I found that the radius of the Blizzard was about 14.8-14.9 yards.

This is what we want *imho* (distance between center of the blizzard and the bunny NPC = radius of the blizzard = 13 yards):
![bunny2](https://cloud.githubusercontent.com/assets/6612710/12297659/755a1d22-ba0e-11e5-936f-375aaa0c2ece.png)

ariel- commented:
The radius is correct at 13yds. The issue here is the dynamic object is calculating targets incorrectly

Treeston commented:
So the range _should be_ 13 yards, and the range _is listed as_ 13 yards in the DBC, but the core is selecting targets that are within ~15 yards instead?

Treeston commented:
Out of interest, what order of magnitude are the map coordinates in Thorim's room?

chaodhib commented:
@ariel- Yep I suppose you are right. I have very limited knowledge about that part of TC so I can't comment any further.

@Treeston The center of the room is at about (2134.877, -263.118, 419.846). Here is an at scale view from above of the room (I have moved the origin to the center of the room and applied a rotation of +90°):
![thorimroom](https://cloud.githubusercontent.com/assets/6612710/12307519/033e830e-ba40-11e5-995b-be7cf7d574c9.png)
The distance from the center to:
* The wall of the room (that forms a big circle) is about 42.75y
* Thorim's platform is about 25.6y
* The circle drawned by the waypoints of the Bunny NPC (the one controlling the blizzard) is about 19.4 y.

Treeston commented:
@chaodhib at first glance, you probably want to check somewhere around [here](https://github.com/TrinityCore/TrinityCore/blob/3.3.5/src/server/game/Spells/Spell.cpp#L1297)

chaodhib commented:
@Treeston Already did. That part is working fine. Like ariel said, it's linked on how dynamic object is calculating targets. This might be a bug in the core affecting all spells with TARGET_DEST_DYNOBJ_ENEMY as target type and/or SPELL_EFFECT_PERSISTENT_AREA_AURA as spell effect.

Shauren commented:
Ref radius of dynobj auras - not the first time this happens: https://github.com/TrinityCore/TrinityCore/blob/6.x/src/server/scripts/Northrend/IcecrownCitadel/boss_lord_marrowgar.cpp#L579

chaodhib commented:
@Shauren Indeed. Shauren to the rescue. This addition does the trick:
```C++            
bool CheckAreaTarget(Unit* target)
{
    if (target->GetExactDist2d(GetOwner()) > 13.0)
        return false;

    /// @todo: fix this for all dynobj auras
    if (target != GetOwner())
    {
        .....
    }
    return true;
}
```
Outside:
![wowscrnshot_011516_190433](https://cloud.githubusercontent.com/assets/6612710/12361025/e16d940e-bbbb-11e5-922b-ed526dd33042.jpg)
Inside:
![wowscrnshot_011516_190436](https://cloud.githubusercontent.com/assets/6612710/12361051/0a10cebc-bbbc-11e5-9e8f-9b08ab0291f6.jpg)
This needs more tests but it looks good!

chaodhib commented:
The only things left regarding the blizzard are:
* Check the speed of the bunny NPC. Making sure he does one cycle in about 47s (cf timestamp on sirikfoll's WPs).
* Check that the blizzard's behavior is blizzlike near Thorim's platform (in that area, there is less than 13y from the bunny NPC to the wall. The center of the blizzard is therefore put as close to the wall as possible)

Treeston commented:
I wonder if we could get 10 people together to playtest this encounter.

Saben65 commented:
Maybe suggest a date and time? I should be free to help.

chaodhib commented:
I'm in

chaodhib commented:
Found the root cause of the problem with the target calculations:

![blizzard bug cause](https://cloud.githubusercontent.com/assets/6612710/12369695/b2fde52e-bbfc-11e5-8a36-681369bd2f4b.png)

Look at the value of maxdist. We are in a particular case where the size of the object (a player in this case) shouldn't be used in the calculation of the distance.

Nikectory commented:
Is the script completed?
When will it be merged?

Aokromes commented:
https://github.com/TrinityCore/TrinityCore/pull/15008#issuecomment-135042600

atlozada commented:
Is there a commit to test if script working? Still cannot activate the fight by killing creatures in the arena. Script missing.

ariel- commented:
@atlozada: needs update after ScriptLoader.cpp changes

This branch has conflicts that must be resolved

Treeston commented:
So @joschiwald what is the current status of this PR?

joschiwald commented:
i'll merge it when i get some freetime

jackpoz commented on src/server/game/Entities/GameObject/GameObject.cpp:
```diff
@@ -638,8 +638,6 @@ void GameObject::Update(uint32 diff)
                 return;
             }
 
-            SetLootState(GO_READY);
```
what is this change about ? shouldn't it be in a separated PR ?

ariel- replied:
Had to move SetLootState *after* flags reset, so the script can be properly notified.

Without this change, `SetLootState` executed the SAI, then the template flags were applied, leaving you with a non usable lever for the rest of the fight.

jackpoz commented on src/server/scripts/Northrend/Ulduar/Ulduar/boss_thorim.cpp:
```diff
@@ -15,29 +15,404 @@
+
+        bool Execute(uint64 /*eventTime*/, uint32 /*updateTime*/) override
+        {
+            _owner->CastSpell((Unit*)nullptr, SPELL_RUNIC_SMASH);
```
is there a better way than using a C-style cast ? static_cast is that bad ?

ariel- replied:
Readability mostly, there are many casts like this, it's the equivalent of CreatureAI::DoCastAOE, since this spell has implicit targets and discards any explicit ones at prepare.

mik1893 commented:
i've tested it yesterday for fun... gg @joschiwald , it's amazing

chaodhib commented on sql/updates/world/3.3.5/2016_xx_xx_xx_world_thorim.sql:
```diff
@@ -0,0 +1,833 @@
+INSERT INTO `creature_addon` (`guid`,`path_id`,`mount`,`bytes1`,`bytes2`,`emote`,`auras`) VALUES (@NPC,@PATH,0,0,1,0, '');
+DELETE FROM `waypoint_data` WHERE `id`=@PATH;
+INSERT INTO `waypoint_data` (`id`,`point`,`position_x`,`position_y`,`position_z`,`orientation`,`delay`,`move_type`,`action`,`action_chance`,`wpguid`) VALUES
+(@PATH,1,2125.243,-279.0981,419.6088,0,0,0,0,100,0),
```
these sniffed waypoints still needs to be cleaned up. Like said above, some points are missing a in a few cycles. 

This should be replaced by  a single rotation.

obsidean commented:
well, i tested this and from an aesthetic perspective, it looks fine.  I am not 100% certain that it is completely Blizzard like, but it is way better than what is there now.  Hopefully someone will push this soon.
Great job by the way.

ariel- commented:
@obsidean: can you point me specifically what is "not completely Blizzard like", I'd like to polish this as much as I can.

obsidean commented:
I am probably missing something,  but would you mind confirming if the chain lightning is giving the correct damage?  
Blizzard Says:
"Strikes an enemy with a blast of lightning that inflicts 3700 to 4300 Nature damage (4625 to 5375 on heroic). The bolt will arc to additional nearby enemies, inflicting 50% more Nature damage to each successive target."

ariel- commented:
Difficulty spells for 10 and 25 man are inserted into spelldifficulty_dbc table, so yes, the right damage spells are chosen according to difficulty.

obsidean commented:
ok, thanks.  from my perspective it looked like the 10 man raid was giving higher damage than the 25 man raid, but I am probably miscalculating.  I am using a damage counter.

Swode commented:
O.o anyone can teach how to fix this? 
code which I have to add?

Help pls :baby:

Treeston commented:
@Swode pull the main 3.3.5 branch, then add `https://github.com/joschiwald/TrinityCore.git` as a new remote (check your git client on how to do that) and merge the thorim branch from that into the main 3.3.5 branch - then compile the result.

obsidean commented:
Does not compile with Mac OS X El Capitan from what I can tell.  I am trying to find what the compile error is exactly.  All I get is 

The following build commands failed:
	CompileC /Users/matteobuonvino/trinity/trinitycore/build/src/server/scripts/TrinityCore.build/Release/scripts.build/Objects-normal/x86_64/boss_thorim.o src/server/scripts/Northrend/Ulduar/Ulduar/boss_thorim.cpp normal x86_64 c++ com.apple.compilers.llvm.clang.1_0.compiler

Swode commented:
http://imgur.com/a/AZoRb

obsidean commented:
You are thinking this is the reason for the compile fail?

ariel- commented:
Needs to be rebased first.

Swode commented:
If I delete the lines that they got me problems I can compile. 
```cpp
if (_hardMode)
				me->SummonGameObject(RAID_MODE(GO_CACHE_OF_STORMS_HARDMODE_10, GO_CACHE_OF_STORMS_HARDMODE_25), 2134.948f, -286.436f, 419.5051f, 1.588249f, 0.f, 0.f, 0.7132502f, 0.7009096f, 604800);
			else
				me->SummonGameObject(RAID_MODE(GO_CACHE_OF_STORMS_10, GO_CACHE_OF_STORMS_25), 2134.948f, -286.436f, 419.5051f, 1.588249f, 0.f, 0.f, 0.7132502f, 0.7009096f, 604800);
```

But I see some bugs for example Sif sometimes dissappear, thorim don't jump down.
When I kill they don't got me the chest...

ariel- commented:
If you delete those lines you won't have any loot. also harpoons are not a part of this encounter.

And this PR hasn't even been merged so it's not a TC repo issue.

obsidean commented:
Confirmed.  did the compile in Xcode and got the same errors as Swode.  Now I remember why I was able to compile previously.  I too had removed those 2 lines from the code.

Treeston commented:
```cpp
if (_hardMode)
                me->SummonGameObject(RAID_MODE(GO_CACHE_OF_STORMS_HARDMODE_10, GO_CACHE_OF_STORMS_HARDMODE_25), Position(2134.948f, -286.436f, 419.5051f), G3D::Quad(1.588249f, 0.f, 0.f, 0.7132502f, 0.7009096f), 604800);
            else
                me->SummonGameObject(RAID_MODE(GO_CACHE_OF_STORMS_10, GO_CACHE_OF_STORMS_25), Position(2134.948f, -286.436f, 419.5051f), G3D::Quad(1.588249f, 0.f, 0.f, 0.7132502f, 0.7009096f), 604800);
```

obsidean commented:
Recompiled with these changes and get the following error:
No member "Quad" in namespace "G3D"

Treeston commented:
`Quat` instead of `Quad`, my bad.

obsidean commented:
Awesome thanks for your prompt response.

obsidean commented:
lol.  sorry for being a nuisance, now getting error No matching constructor for "G3D:Quat"

Treeston commented:
Yeah I messed up the brackets.

```cpp
if (_hardMode)
                me->SummonGameObject(RAID_MODE(GO_CACHE_OF_STORMS_HARDMODE_10, GO_CACHE_OF_STORMS_HARDMODE_25), Position(2134.948f, -286.436f, 419.5051f, 1.588249f), G3D::Quat(0.f, 0.f, 0.7132502f, 0.7009096f), 604800);
            else
                me->SummonGameObject(RAID_MODE(GO_CACHE_OF_STORMS_10, GO_CACHE_OF_STORMS_25), Position(2134.948f, -286.436f, 419.5051f, 1.588249f), G3D::Quat(0.f, 0.f, 0.7132502f, 0.7009096f), 604800);
```

obsidean commented:
tks again.

obsidean commented:
Swode: About harpoons, that is for the dragon boss not Thorim.

obsidean commented:
Seems that Thorim doesn't jump off the platform anymore, for whatever reason.

obsidean commented:
Should I create a ticket for this or does it stay under this one?

Treeston commented:
PR feedback should be in PR comments.

Issues are for problems with the main branch(es).

obsidean commented:
Here is an interesting video that may be handy
https://www.youtube.com/watch?v=KD7a06Lkrwc
For one thing it shows that the gate stays opened for 59sec.  Currently it does not.

obsidean commented:
If I wanted to remove that spell which sends that ball of lighting which one shots the raid, which part of the code controls that?

Treeston commented:
Check the events enum and take a wild guess which one you're looking for.

PS: We're not your support staff for hacking a work in progress PR.

joschiwald commented:
take a look at the arena npc enterevademode

obsidean commented:
ok fair enough.  Sorry I asked

obsidean commented:
so here is what I discovered in regards to Thorim not jumping off platform:
(i) If you stay on the platform too long while having engaged Thorim, he will not jump;i.e. he will fight you on the platform
(ii) IF you do then jump off, he'll follow you, but be in EVADE mode.

Hope this helps

ariel- commented:
> Here is an interesting video that may be handy
> https://www.youtube.com/watch?v=KD7a06Lkrwc
> For one thing it shows that the gate stays opened for 59sec. Currently it does not.

Time for closing the door is a WDB field in the gameobject_template table, this comes from sniffs and cannot be modified. Perhaps it was changed on a later expansion; but the current value of 8 sec is blizzlike

Treeston commented:
What todos are remaining before merge? @ariel-?

ariel- commented:
Summoning @joschiwald

Keader commented:
Nothing yet?

Keader commented:
Nothing yet?

ccrs commented:
probably the void ````JustRespawned() override```` should be deleted now after Treeston's commit

Keader commented on src/server/scripts/Northrend/Ulduar/Ulduar/boss_thorim.cpp:
```diff
@@ -15,29 +15,403 @@
+    SAY_AGGRO_2                                 = 1,
+    //SAY_SPECIAL_1                               = 2,
+    SAY_SPECIAL_2                               = 3,
+    //SAY_SPECIAL_3                               = 4,
```
?

Keader commented on src/server/scripts/Northrend/Ulduar/Ulduar/boss_thorim.cpp:
```diff
@@ -66,27 +522,397 @@ class boss_thorim : public CreatureScript
+
+                if (_hardMode)
+                {
+                    if (Creature* Sif = ObjectAccessor::GetCreature(*me, instance->GetGuidData(DATA_SIF)))
```
sif

Keader commented on src/server/scripts/Northrend/Ulduar/Ulduar/boss_thorim.cpp:
```diff
@@ -66,27 +522,397 @@ class boss_thorim : public CreatureScript
+
+                events.SetPhase(PHASE_1);
+
+                events.ScheduleEvent(EVENT_SAY_AGGRO_2, 9000, 0, PHASE_1);
```
All events can be changed by std::chrono, is more readable i think ^^

Keader commented on src/server/scripts/Northrend/Ulduar/Ulduar/boss_thorim.cpp:
```diff
@@ -66,27 +522,397 @@ class boss_thorim : public CreatureScript
+                            break;
+                        case EVENT_UNBALANCING_STRIKE:
+                            DoCastVictim(SPELL_UNBALANCING_STRIKE);
+                            events.ScheduleEvent(EVENT_UNBALANCING_STRIKE, urand(15000, 20000), 0, PHASE_2);
```
Since https://github.com/TrinityCore/TrinityCore/commit/2174c5d7dd9669384b2e68b8b406a3173042d7d9 
You can simplify to events.ScheduleEvent(EVENT_UNBALANCING_STRIKE, Seconds(8), Seconds(10)), for "random times"

Keader commented on src/server/scripts/Northrend/Ulduar/Ulduar/boss_thorim.cpp:
```diff
@@ -66,27 +522,397 @@ class boss_thorim : public CreatureScript
+                            break;
+                        case EVENT_SUMMON_ADDS:
+                            SummonWave();
+                            events.ScheduleEvent(EVENT_SUMMON_ADDS, _orbSummoned ? 3000 : 10000, 0, PHASE_1);
```
why not events.Repeat()?

Treeston commented:
@Keader all the readability overloads for `std::chrono` happened after this PR was originally authored, they can be added as final clean-up pass (once @joschiwald is confident in implementation)

Keader commented on src/server/scripts/Northrend/Ulduar/Ulduar/boss_thorim.cpp:
```diff
@@ -66,27 +522,397 @@ class boss_thorim : public CreatureScript
+                        {
+                            // Event starts
+                            me->RemoveFlag(UNIT_FIELD_FLAGS, UNIT_FLAG_IMMUNE_TO_PC);
+                            DoZoneInCombat(me, 250.0f);
```
250.0f is new default value for DoZoneInCombat, check : https://github.com/TrinityCore/TrinityCore/blob/3.3.5/src/server/game/AI/CreatureAI.cpp#L49

ariel- commented:
What commit?

2016-10-20 12:31 GMT-03:00 ccrs <notifications@github.com>:

> probably the void JustRespawned() override should be deleted now after
> Treeston's commit
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/TrinityCore/TrinityCore/pull/15008#issuecomment-255140735>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AInQJ3DXUeJsZ8bmpz4FiWnLXQfU2pN-ks5q14k4gaJpZM4FR8-U>
> .
>

ccrs commented:
aab7abfd045b935e112feec4b192fb7921394766

Keader commented on src/server/scripts/Northrend/Ulduar/Ulduar/boss_thorim.cpp:
```diff
@@ -15,29 +15,403 @@
+                    return false;
+                case 1:
+                    _owner->SetReactState(REACT_AGGRESSIVE);
+                    _owner->AI()->DoZoneInCombat(_owner, 200.0f);
```
new default value is 250.0f

Viste commented:
hope it will be merge

chaodhib commented:
There are some issues with the SQL script FYI (GUID conflict in gameobject table)

Aokromes commented:
things of being one over year and half old pr :P

chaodhib commented:
Yeah yeah I know :p I just meant that the PR needs a slight update. The rest works very well.

Keader commented:
Needs a merge, just saying... (rolleyes)

joschiwald committed:
Scripts/Ulduar: Initial implementation of the Thorim Encounter

Closes #7651

Fix logic fail in achievement:
SPELL_LIGHTNING_CHARGE aka 62279 is casted on Thorim itself to buff him, not the damage spell tracked by the achievement.

ariel- committed:
Core/Scripts: improvements on the Thorim script

Improvements done to initial work by joschiwald:

- Fixed stormhammer casting and visuals
- Fixed spell credits
- Implemented removal of Impale when HP is higher than 90%
- Corrected faction templates for the pre combat creatures
- Implemeted Leap on the arena adds (thanks to joschiwald for implementing conditions)
- Scripted lighting charge using AuraScript periodic
- Fixed timings
- Corrections on the hallway encounters (added knockback immunity to minibosses too)
- Fixed multiple blizzards issue
- Lever will now reset properly, should players fail to get inside. This allows the door to be opened more than once during the combat
- Corrections in the outro event
- Implemented paralytic field traps on the hallway.
- Implemented Ancient Gate of the Keepers opening. Fixes #17072

Closes #15008

ariel- commented:
Updated. Removed radius check due to #16290 being merged

chaodhib commented:
I still get hit in the safe point described at https://github.com/TrinityCore/TrinityCore/pull/15008#issuecomment-143545353. I suspect the waypoints:
https://cloud.githubusercontent.com/assets/6612710/24274393/28cf25ee-1028-11e7-8453-981e46604b98.png

That safe point is at 18 yards from the center. If we change the path taken by the bunny NPC to be a **perfect** circle of 19.5y radius (and because the hitbox size of a player is 1.5y), it will work. And this radius fits well (with an error of 0.1y) [the observations I made ](https://github.com/TrinityCore/TrinityCore/pull/15008#issuecomment-170361636)on the bunny NPC path.

In conclusion, I'm pretty sure Blizzard does not handle the path of this NPC with static way-points. Instead, I think they use a "Circle Movement Generator". This movement generator make the NPC moves on an accurate circle on the back end side. And only (inaccurate) way-points are sent to the client (for the visual of the blizzard).

ariel- commented:
I cleaned them pretty much after noticing there was a destroy object. You didn't get hit after clean waypoints and radius check in aura. Maybe still targets too far away?.

BTW, we have a MoveCirclePath in MotionMaster

chaodhib commented:
@ariel- 
The custom radius check was too lenient. It didn't account for the hitbox of the player, when it should have. So it was a 1.5 yards bonus given for us. The incorrect waypoints of the bunny NPC didn't matter then. Now that we are using the correct formula, we have to more precise.

Here are the current way-points of the bunny NPC:

![image](https://cloud.githubusercontent.com/assets/6612710/24331799/f44b783e-123b-11e7-9f8d-c01822eddcb1.png)

There is clearly something wrong with them.

Now, using a sniff of Ulduar (and [SniffExplorer](https://github.com/chaodhib/SniffExplorer)), I filtered the ON_MONSTER_MOVE packets of the bunny NPC controlling the motion of the blizzard: https://gist.github.com/chaodhib/4c8fae817b10abdfcd23f391fd0a2697#file-thorim-blizzard-txt

In it, the final point of the packets repeat themselves after 32 packets. In other words, the NPC is always going through the same 32 points. And these 32 points describe a cycle:
```
X=2148.425, Y=-276.7478, Z=419.5923
X=2145.229, Y=-278.5732, Z=419.6016
X=2142.105, Y=-280.7194, Z=419.5932
X=2138.419, Y=-281.2838, Z=419.5999
X=2134.843, Y=-281.9885, Z=419.5996
X=2131.128, Y=-281.278, Z=419.5985
X=2127.415, Y=-280.8166, Z=419.5875
X=2124.459, Y=-278.4777, Z=419.6088
X=2121.098, Y=-276.7148, Z=419.5924
X=2119.281, Y=-273.4217, Z=419.6069
X=2116.933, Y=-270.4605, Z=419.5851
X=2116.359, Y=-266.7142, Z=419.6042
X=2115.436, Y=-263.0562, Z=419.5847
X=2116.543, Y=-259.411, Z=419.6082
X=2116.943, Y=-255.605, Z=419.5851
X=2119.377, Y=-252.7493, Z=419.6096
X=2121.183, Y=-249.3546, Z=419.5858
X=2124.49, Y=-247.5925, Z=419.6089
X=2127.329, Y=-245.1714, Z=419.5925
X=2131.097, Y=-244.8609, Z=419.6089
X=2134.792, Y=-243.7036, Z=419.585
X=2138.466, Y=-244.3162, Z=419.5979
X=2142.386, Y=-244.7266, Z=419.5746
X=2145.455, Y=-247.1502, Z=419.5968
X=2148.564, Y=-249.0781, Z=419.5791
X=2150.63, Y=-252.4069, Z=419.598
X=2153.104, Y=-255.5907, Z=419.5758
X=2153.286, Y=-259.4698, Z=419.6053
X=2154.193, Y=-263.0425, Z=419.6039
X=2153.279, Y=-266.768, Z=419.6026
X=2152.758, Y=-270.5462, Z=419.5822
X=2150.38, Y=-273.4614, Z=419.605
```

Here is a representation of the points:
![image](https://cloud.githubusercontent.com/assets/6612710/24331883/5de90c06-123d-11e7-9f00-6e7191a7e3a0.png)

And here is the SQL to use those WPs:
```
-- Pathing for Thorim Event Bunny Entry: 32892 'TDB FORMAT'
SET @NPC := 136515;
SET @PATH := @NPC * 10;
UPDATE `creature` SET `spawndist`=0,`MovementType`=2,`position_x`=2138.33,`position_y`=-281.1643,`position_z`=419.8471 WHERE `guid`=@NPC;
DELETE FROM `creature_addon` WHERE `guid`=@NPC;
INSERT INTO `creature_addon` (`guid`,`path_id`,`mount`,`bytes1`,`bytes2`,`emote`,`auras`) VALUES (@NPC,@PATH,0,0,1,0, '');
DELETE FROM `waypoint_data` WHERE `id`=@PATH;
INSERT INTO `waypoint_data` (`id`,`point`,`position_x`,`position_y`,`position_z`,`orientation`,`delay`,`move_type`,`action`,`action_chance`,`wpguid`) VALUES
(@PATH,1,2148.425,-276.7478,419.5923,0,0,0,0,100,0),
(@PATH,2,2145.229,-278.5732,419.6016,0,0,0,0,100,0),
(@PATH,3,2142.105,-280.7194,419.5932,0,0,0,0,100,0),
(@PATH,4,2138.419,-281.2838,419.5999,0,0,0,0,100,0),
(@PATH,5,2134.843,-281.9885,419.5996,0,0,0,0,100,0),
(@PATH,6,2131.128,-281.278,419.5985,0,0,0,0,100,0),
(@PATH,7,2127.415,-280.8166,419.5875,0,0,0,0,100,0),
(@PATH,8,2124.459,-278.4777,419.6088,0,0,0,0,100,0),
(@PATH,9,2121.098,-276.7148,419.5924,0,0,0,0,100,0),
(@PATH,10,2119.281,-273.4217,419.6069,0,0,0,0,100,0),
(@PATH,11,2116.933,-270.4605,419.5851,0,0,0,0,100,0),
(@PATH,12,2116.359,-266.7142,419.6042,0,0,0,0,100,0),
(@PATH,13,2115.436,-263.0562,419.5847,0,0,0,0,100,0),
(@PATH,14,2116.543,-259.411,419.6082,0,0,0,0,100,0),
(@PATH,15,2116.943,-255.605,419.5851,0,0,0,0,100,0),
(@PATH,16,2119.377,-252.7493,419.6096,0,0,0,0,100,0),
(@PATH,17,2121.183,-249.3546,419.5858,0,0,0,0,100,0),
(@PATH,18,2124.49,-247.5925,419.6089,0,0,0,0,100,0),
(@PATH,19,2127.329,-245.1714,419.5925,0,0,0,0,100,0),
(@PATH,20,2131.097,-244.8609,419.6089,0,0,0,0,100,0),
(@PATH,21,2134.792,-243.7036,419.585,0,0,0,0,100,0),
(@PATH,22,2138.466,-244.3162,419.5979,0,0,0,0,100,0),
(@PATH,23,2142.386,-244.7266,419.5746,0,0,0,0,100,0),
(@PATH,24,2145.455,-247.1502,419.5968,0,0,0,0,100,0),
(@PATH,25,2148.564,-249.0781,419.5791,0,0,0,0,100,0),
(@PATH,26,2150.63,-252.4069,419.598,0,0,0,0,100,0),
(@PATH,27,2153.104,-255.5907,419.5758,0,0,0,0,100,0),
(@PATH,28,2153.286,-259.4698,419.6053,0,0,0,0,100,0),
(@PATH,29,2154.193,-263.0425,419.6039,0,0,0,0,100,0),
(@PATH,30,2153.279,-266.768,419.6026,0,0,0,0,100,0),
(@PATH,31,2152.758,-270.5462,419.5822,0,0,0,0,100,0),
(@PATH,32,2150.38,-273.4614,419.605,0,0,0,0,100,0);
```

I tested them and it looks good. The safe points tested are indeed safe.

chaodhib commented:
I confirm. The 32 hard-coded points mentioned above are indeed way-points. In order to prove that:

I plotted the 32 hardcoded points: 
![plot no splines](https://cloud.githubusercontent.com/assets/6612710/24378545/99c7fe7e-1343-11e7-8995-b691e31b16c5.jpg)

Then, on the same graph, I also plotted the way-points of the first 32 packets found in my sniff (cf above):
![plot 32 splines](https://cloud.githubusercontent.com/assets/6612710/24378548/9f33ca46-1343-11e7-9ebf-89996dd13152.jpg)

And then the waypoints of every packets:
![plot every splines](https://cloud.githubusercontent.com/assets/6612710/24378566/b00cad06-1343-11e7-8719-3818f8826f46.jpg)

In each packet, there is a spline. From the plots, we can see that the points of each spline are always between 2 of the 32 hardcoded points. So, it's pretty clear that a creature following those ON_MONSTER_MOVE packets is just moving between the 32 hardcoded points.

So in conclusion, unless I made a mistake somewhere, the 32 points mentioned above can indeed be seen as waypoints. Therefore the SQL script provided above should conclude Sif's Blizzard issue. Few finally. 

PS: I'm pretty sure waypoints are handled differently in TC than in retail (if i'm not mistaken, TC sends one point at a time. While in retail, a creature has more or less enough data to move toward the next 2 waypoints). But that's beside the point.

ariel- commented:
Excellent, will update later. Thanks for your explanations and graphs

Keader commented on sql/updates/world/3.3.5/2017_MM_DD_NN_world_15008.sql:
```diff
@@ -0,0 +1,324 @@
+UPDATE `gameobject_template` SET `size`=3 WHERE `entry`=194315;
+UPDATE `gameobject_template_addon` SET `faction`=94, `flags`=16 WHERE `entry` IN (194313,194314,194315);
+
+SET @OGUID := 3926; -- 4 free guids
```
this guid is not free anymore xd

ariel- replied:
SQL applied without errors

ariel- committed:
Remove IsInArena, replace by boundaries system

Kittnz commented:
"But that's beyond the scope of this PR."
I smell a new PR coming! Good job!

ariel- committed:
Update blizzard

- Changed blizzard bunny targetting to conditions entirely.
- Updated bunny waypoints thanks to chaodhib
- Removed obsolete scripts
- Updated conditions using Malcrom's Condition Creator
- Delete spell script names by ScriptName instead of spell_id

ariel- committed:
Removed outdated comment

ariel- commented:
:shipit:

Aokromes commented:
push the shinny button!

ccrs commented on src/server/scripts/Northrend/Ulduar/Ulduar/ulduar.h:
```diff
@@ -425,7 +425,7 @@ enum UlduarData
     DATA_HODIR_YS,
     DATA_THORIM_YS,
     DATA_MIMIRON_YS,
-    DATA_ILLUSION, // = 38 used by conditions
```
why outdated?

ariel- replied:
It was moved to cpp a long time ago. And tbh it wasn't even in the scope of the PR

ccrs commented on src/server/scripts/Northrend/Ulduar/Ulduar/boss_thorim.cpp:
```diff
@@ -95,7 +904,1269 @@ class boss_thorim : public CreatureScript
+            bool CanAIAttack(Unit const* who) const override
+            {
+                // don't try to attack players in balcony
+                if (_isInArena && who->GetPositionZ() > 428.0f)
```
use boundaries here?

ariel- replied:
I tried, funny things happen with ZRange boundary (you need to specify both a min and a max, and I a min only)

ccrs commented on src/server/scripts/Northrend/Ulduar/Ulduar/boss_thorim.cpp:
```diff
@@ -95,7 +904,1269 @@ class boss_thorim : public CreatureScript
+        }
+};
+
+struct npc_thorim_minibossAI : public ScriptedAI
```
this script has no conditions for the unit to not attack arena players (might get aggro from healing or similar)

ariel- replied:
Yeah, forgot to add boundaries to them when revamping

ccrs commented on src/server/scripts/Northrend/Ulduar/Ulduar/boss_thorim.cpp:
```diff
@@ -95,7 +904,1269 @@ class boss_thorim : public CreatureScript
+                        && me->IsWithinLOSInMap(target))
+                        allowMove = false;
+                }
+                SetCombatMovement(allowMove);
```
this wont do nothing once the combat is started iirc

ccrs replied:
correction, will work since there is currently an infinite loop in UpdateVictim() (on each update tick a new movement is created), so better do this correctly since that will be fixed (soon 💃 )

ariel- replied:
Will have to review this part of the code (AI code of the trash is almost intact work of joschiwald)

ccrs commented on src/server/scripts/Northrend/Ulduar/Ulduar/boss_thorim.cpp:
```diff
@@ -66,27 +505,397 @@ class boss_thorim : public CreatureScript
+                            me->RemoveAurasDueToSpell(SPELL_SHEATH_OF_LIGHTNING);
+                            me->SetReactState(REACT_AGGRESSIVE);
+                            SetCombatMovement(true);
+                            me->GetMotionMaster()->MoveJump(2134.79f, -263.03f, 419.84f, me->GetOrientation(), 30.0f, 20.0f);
```
better create the Chase movement on MovementInform (jump movement is finished)

ccrs replied:
and better reset its aggro table, so it doesnt insta target healers, right? :)

ariel- replied:
Healers are meant to die

ccrs commented on src/server/scripts/Northrend/Ulduar/Ulduar/boss_thorim.cpp:
```diff
@@ -66,27 +505,397 @@ class boss_thorim : public CreatureScript
+
+                        Movement::MoveSplineInit init(summon);
+                        init.MovebyPath(path);
+                        init.Launch();
```
uuuuuuuuuuuugh

better use treeston's spline generator than this

ariel- replied:
Don't have sniff for electric ball of DOOM (tm), the waypoints are joschiwald's work

ccrs commented on src/server/scripts/Northrend/Ulduar/Ulduar/boss_thorim.cpp:
```diff
@@ -66,27 +505,397 @@ class boss_thorim : public CreatureScript
+                me->SetReactState(REACT_PASSIVE);
+                me->InterruptNonMeleeSpells(true);
+                me->RemoveAllAttackers();
+                me->AttackStop();
```
isnt there a method with this calls already? ... CombatStop()? ugh my memory D:

ariel- replied:
It's different. We mustn't stop combat, rather keep on combat until the outro finishes

ccrs commented on src/server/scripts/Northrend/Ulduar/Ulduar/instance_ulduar.cpp:
```diff
@@ -648,11 +686,23 @@ class instance_ulduar : public InstanceMapScript
+                                }
+                            }
 
                             instance->SummonCreature(NPC_THORIM_OBSERVATION_RING, ObservationRingKeepersPos[2]);
```
something should be done in the future about this, once the grid unloads the instance doesnt resummon em on reload

ariel- replied:
Pending dynamic spawning I guess

ccrs commented on src/server/scripts/Northrend/Ulduar/Ulduar/boss_thorim.cpp:
```diff
@@ -95,7 +904,1269 @@ class boss_thorim : public CreatureScript
+                if (SpellInfo const* spellInfo = sSpellMgr->GetSpellInfo(spellId))
+                {
+                    if (me->IsInRange(target, spellInfo->GetMinRange(false), spellInfo->GetMaxRange(false))
+                        && me->IsWithinLOSInMap(target))
```
Chase movement using the spell range as offset already does this, seems like overkill to me

ariel- replied:
Yeah, I don't see the purpose of it either

ccrs commented on src/server/scripts/Northrend/Ulduar/Ulduar/boss_thorim.cpp:
```diff
@@ -95,7 +904,1269 @@ class boss_thorim : public CreatureScript
+            return target;
+        }
+
+        static Unit* GetHealTarget(uint32 spellId, Unit* caster)
```
Shouldnt this be the general behaviour of "gethealtarget" methods? I mean, shouldnt this be done to correct the global searchers? There are a couple scripts that would benefict from this.

ariel- replied:
There's a coupling in that case. Searchers are just gimme any target in range. AI then decides what to do with those targets.

IMO Searchers are good as they are

Keader commented:
Tested today, in 25 man
Tested in a clean Trinity and Clean DB in rev 21b2042840e05ebeba365c54d47c07c13fde123f

**Issue 1:**
Idk if is caused by this PR, but Acient Gate default state is open now.
Print with all Keepers alive:
<details>

![image](https://cloud.githubusercontent.com/assets/7120913/24910104/049615ce-1e9d-11e7-9be5-9f59bfe0998f.png)
</details>

He should be closed. When all keepers die will unlock and will open when players click.

**Issue 2:**
If i remember right, Minions in Thorim Room should be in combat with Jormungar...
In current state of this PR, all minions is out of combat.
Combat start only if i pull Jormungar.

-- Edit

After wipe, some creatures enter in combat, but the iron soldier remain out of combat

**Issue 3:**

Inside of tunnel, Npc Runic Colossus (entry 32872) cant be attacked.
Return "INVALID TARGET" message

**Issue 4:**

With 1 character in main room, and 1 character in tunnel...
Character in tunnel got HK by a Energy Ball.
If i remember right, it should happens only if all players in main room die.
(maybe caused by boundary?)

**Issue 5:**

Try taunt thorim while he's talking:

<details>

Thorim yells: Interlopers! You mortals who dare to interfere with my sport will pay.... Wait--you...
Thorim yells: I remember you... in the mountains.... But you... what is this? Where am--

<!details>

will cause HK lol

ariel- committed:
got feedback

- swapped factions (alliance should get horde pre adds and vs)
- refined boundaries
- refactored code
- fixed sethomeposition leap script
- removed a bunch of magic numbers
- runic colossus should finish current runic explosion before beginning to attack
- fixed updateAI logic to put it in line with other scripts

ariel- commented:
> Idk if is caused by this PR, but Acient Gate default state is open now.
Print with all Keepers alive:

fixed

> If i remember right, Minions in Thorim Room should be in combat with Jormungar...
In current state of this PR, all minions is out of combat.
Combat start only if i pull Jormungar.

maybe because of grid activation?

> Inside of tunnel, Npc Runic Colossus (entry 32872) cant be attacked.
Return "INVALID TARGET" message

core bug, packet doesn't get broadcast at sufficient distance, because he doesn't have flags set

> With 1 character in main room, and 1 character in tunnel...
Character in tunnel got HK by a Energy Ball.
If i remember right, it should happens only if all players in main room die.
(maybe caused by boundary?)

should be fixed

> Try taunt thorim while he's talking:

can't taunt, balcony blocks line of sight

Treeston commented:
* Don't rely on LoS blocks, just handle it gracefully.
*  Creatures on the grid can be loaded in any arbitrary order. Check on load for any of them whether all others are present, then start the event. Don't rely on any specific creature being loaded last.

ariel- commented:
I tried taunting using .gm fly, no real issues here, so it must've been fixed by the boundaries tuning

Keader commented:
Tested today:

Gate fixed
Energy Ball Hit Kill Fixed

> maybe because of grid activation?
Idk
But is working after first wipe, only Dark Rune Acolyte (entry=32886) dont attack Jormungar.

Taunt thing i cant reproduce anymore, i realy dont know how i can taunt in main room o.o
btw, testing with .gm fly on, is working, i dont get hit kill anymore
So... left only Jormungar and trash mobs bug, related with this pr

obsidean commented:
I received the following build error:
no member named 'setFaction' in 'Creature'
                me->setFaction(FACTION_FRIENDLY);
